### ISSUE TITLE
Multiple UPE - Spike: Explore scope of creating new feature flag for Split-UPE

### ISSUE THREAD

**Issue by FangedParakeet** [Automattic/woocommerce-payments#5271]
### Description

Following #4671 it has been decided that the best plan to rollout the first phase of the split-UPE project would be to create a new feature flag to be used in parallel with the existing UPE implementation. Downgrading the UPE for merchants would remove the additional payment methods from merchant checkouts, which would be unacceptable since these payment methods receive considerable usage. Moreover, we would prefer to gradually phase out the existing UPE implementation, to ensure that the split-UPE does not create any major conflicts in production, so replacing this existing UPE code entirely--as has been the current design--would also not be agreeable.

Consequently, our confirmed rollout plan must initially support the non-UPE gateway, legacy UPE gateway, and split-UPE gateway simultaneously. Once the final phase of the split-UPE has been implemented and released without incident, supporting [Stripe's new deferred intent payment flow](https://stripe.com/docs/payments/defer-intent-creation) and supporting compatibility with WooPay, we can plot to remove the redundant gateways and unify the payment gateways into one. 

Since our current POC branch #3968 unfortunately overwrites the previous UPE implementation, we will now have to work backwards somewhat to ensure that we can reintegrate the previous implementation and separate it from the split-UPE with a new feature flag. This issue is to try to scope out what would be required to create this separation and how much effort it would cost our squad. 

### Additional context

#4671 
pc2DNy-2qG-p2#comment-3837

**Comment by FangedParakeet** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/Heisenberg-63586f2ec60359001310fa6a/issues/Automattic/woocommerce-payments/5271?planning-poker) with Zenhub @mdmoore @timur27

**Comment by FangedParakeet** [comment]
Scanning through #3968, I've basically grouped the additions to the UPE made by the split-UPE implementation into the following categories.

- Multiple payment gateways registered to plugin, instead of single.
- Multiple payment elements mounted/registered to the shortcode/blocks checkout, instead of single element.
- Multiple payment intent caches for each payment method type, instead of single payment intent cache.
- Multiple AJAX functions to handle intent actions (create setup intent, create payment intent, update intent) for each payment method registered as hooks, instead of single hook.
- Testing instructions for payment gateway is dependent on the payment method type, instead of always displaying card testing instructions.
- WC Payment tokens are stored with separate gateway IDs, dependent on the payment method type, instead of always being the standard CC/`woocommerce_payments` gateway ID.

These changes are not actually too drastic in the back-end of the plugin; however, the front-end scripts have the largest differences. 

I think if we create a new gateway class for the split-UPE, we can minimise the efforts required here. The new split-UPE gateway class can inherit from `UPE_Payment_Gateway`, which in turn inherits from `WC_Payment_Gateway_WCPay`, while adding our own split-UPE implementation idiosyncrasies inside this new child class. This way we can create new scripts files (similar to how `/client/checkout/classic/upe.js` and `/client/checkout/blocks/upe.js` exist separate from their `index.js` compatriots) that we can decide to enqueue depending on which gateway class is being used. Similarly the hooks will be added only when the respective gateway class is present. 

The ugliest branching logic will likely be present inside the `WC_Payments_Token_Service` class and the `WC_Payments` class, where the gateways are actually registered, but hopefully we can minimise the rest of the conflicts by keeping the rest of the logic constrained in separate gateway classes.

I have created #5353 to continue this grand vision as described above. I think this is the best path forward, but I'd love to hear the feedback on this approach from my squadron comrades, @mdmoore & @timur27, so please shower me with your wisdom now or forever hold your peace. Cheers!

**Comment by timur27** [comment]
> I think if we create a new gateway class for the split-UPE, we can minimise the efforts required here. The new split-UPE gateway class can inherit from UPE_Payment_Gateway, which in turn inherits from WC_Payment_Gateway_WCPay, while adding our own split-UPE implementation idiosyncrasies inside this new child class.

Sounds like a good plan at first glance. 👍 

------------

> The ugliest branching logic will likely be present inside the `WC_Payments` class

Agree here.

If I understand it correctly, the following configurations are available. Please correct me in case I missed anything.

| |  `_wcpay_feature_upe` | `_wcpay_feature_split_upe` |
| --- | --- | --- |
| **non-UPE** | 0 | 0 |
| **UPE** | 1 | 0 |
| **split-UPE** | 0 | 1 |
| **split-UPE** | 1 | 1 |


#### **Avoid complex `if` branching in `class-wc-payments`**
With the aforementioned features matrix, if we continue to setup payments the way we do right now, we will have something similar to: 

```php
if ( WC_Payments_Features::is_upe_enabled() ) {
	if ( WC_Payments_Features::is_split_upe_enabled() ) {
		// initialize the system's state with multiple payment gateways
	} else {
		// initialize the system's state with one WC_Payment_Gateway_WCPay
		// reverted code needs to go here

	}
} else {
    // initialize the system's state for non-UPE feature
}
```

This looks scary and probably will be hard to manage in the future. I think that the following steps might make our lives easier either now or in the future. 

1. Let `WC_Payments_Features` answer the question of what should we initialize.

```php
/**
* 0 0 for non-UPE
* 1 0 for UPE
* 0 1 for split-UPE
* 1 1 for split-UPE
*/
public static function get_upe_strategy_to_use() {
	if ( ! self::is_upe_enabled() && ! self::is_split_upe_enabled() ) {
		return 'non_upe';
	}

	if ( self::is_upe_enabled() && ! self::is_split_upe_enabled() ) {
		return 'upe';
	}

	return 'split_upe';
}
```

2. Create a new class, for instance, `WC_Payments_UPE_Factory`.
```php
class WC_Payments_UPE_Factory {
	private $strategies;

	public function __construct($strategies)
	{
		$this->strategies = [
			'non-upe' => $this->non_upe(),
			'split_upe' => $this->split_upe(),
			'upe' => $this->upe(),
		];
	}

        // the logic goes here
        function split_upe() {}
        // here
	function upe() {}
        // and here
	function non_upe() {}
}
```

3. Let the factory initialize the state of the WCPay gateways/payment methods based on the strategy
```php
$factory->get_strategies()[WC_Payments_Features::get_upe_strategy_to_use()]();
```

There are some dependencies to e.g. `$upe_payment_method_map`/`$upe_payment_gateway_map` and some more, which need a more elegant found for them to be handled.

These are just raw ideas straight from the top of my head, so please don't treat them as instructions but rather as brainstorming. I also understand that we might be short on time if we decide to introduce the solution from an ideal world rather than from the world where everything just works. I didn't analyze `WC_Payments_Token_Service` yet.

------------

💡 Not sure how problematic `WC_Payments_UPE_Checkout` and `WC_Payments_Checkout` will be in terms of our refactor as such.

**Comment by FangedParakeet** [comment]
>If I understand it correctly, the following configurations are available. Please correct me in case I missed anything.

Nope, that matrix looks correct, as per my own understanding!

I totally agree that the `WC_Payments_Features` should be used to isolate this logic as much as possible. I also like your idea of using create a Gateway Factory class to encapsulate this logic as well. Presumably that would replace [this lumbering hulk of code](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/class-wc-payments.php#L337-L362). This scope of this refactor will extend a little further beyond the frontiers of our work, but as long as it doesn't cause too many headaches, I think it will be worthwhile. 

The main difference in the `WC_Payments_Token_Service` class is that normally all saved tokens have the common gateway ID of `woocommerce_payments`, whereas with the split-UPE enabled they are now stored under their own payment method/gateway specific gateway IDs. I'm optimistic that if we handle this case and the effects of translating existing tokens while enabling or disabling the UPE, the rest will click together, but I'm expecting this to not end up being all so simple.

The `WC_Payments_UPE_Checkout` class is predominantly there to enqueue scripts and styles and display the payment element container. This is essentially the same, even in the split-UPE branch, with the exception being that the `payment_fields()` function is called repeatedly for each payment gateway and the container itself contains a `data-payment-method-type` attribute to indicate the payment method/gateway. I'm hopeful this won't be too difficult to integrate together.

**Comment by FangedParakeet** [comment]
I think this spike can now be closed, since the implementation will now be completed in #5353 and there aren't really any more questions to be answered here for now.
